dev-tools.man
.\" author: Les Ateliers Pierrot

.TH  "Dev Tools" "3" "2013-10-20" "Version 1.0.7" "dev-tools.sh Manual"

.SH NAME

.PP
dev-tools - Packages development & deployment facilities

.SH SYNOPSIS

.PP
\fBbash-library-script [common options] [script options [=value]] action --\fP

.PP
\fBdev-tools.sh\fP  [\fB-h\fP]  [\fB-f\fP|\fB-i\fP|\fB-q\fP|\fB-v\fP]  [\fB-x\fP|\fB--dry-run\fP]  ... 
    ... [\fB--var\fP \fI=name\fP]  [\fB--val\fP \fI=value\fP]  [\fB--filename\fP]  [\fB--full\fP]  config
    ... [\fB--env\fP \fI=env\fP]  deploy
    ... [\fB--begin\fP \fI=mask\fP]  [\fB--end\fP \fI=mask\fP]  [\fB--output\fP \fI=filename\fP]  [\fB--filename\fP]  extract
    ... [\fB--files\fP \fI=chmod\fP]  [\fB--dirs\fP \fI=chmod\fP]  [\fB--bin\fP \fI=path\fP]  [\fB--bin-mask\fP \fI=mask\fP]  fix-rights
    ... [\fB--source\fP \fI=path\fP]  [\fB--filename\fP \fI=filename\fP]  [\fB--type\fP \fI=type\fP]  [\fB--dir\fP \fI=dir path\fP]  [\fB--markdown\fP \fI=bin path\fP]  [\fB--whatis\fP \fI=bin path\fP]  [\fB--makewhatis\fP \fI=bin path\fP]  manpage
    ... [\fB--env\fP \fI=env\fP]  [\fB--target\fP \fI=server\fP]  [\fB--options\fP \fI="rsync options"\fP]  sync
    ... [\fB--name\fP \fI=version\fP]  [\fB--branch\fP \fI=name\fP]  [\fB--hook\fP \fI=path\fP]  [\fB--no-hook\fP]  version-tag
    \fB-p | --project\fP \fI=path\fP  \fI<action> in\fP \fBcleanup | config | deploy | extract | fix-rights | manpage | sync | version-tag\fP  --

.SH DESCRIPTION

.PP
The `\fSdev-tools\fP` is one single shell script that handles a set of available actions
(defined themselves as shell scripts) to execute something upon a package in development. The usage is
quite simple as it just requires to understand the command line call of one single script.
The global script always follows the same rules and acts like a dispatcher that distributes the
options to an action. More, creating a new action (such as your own actions) is as simple
as writing a new shell script in the `\fSdev-tools-actions/\fP` directory and call it with the global
script.

.PP
Run option `\fS-h action\fP` to see the help about a specific action and use option `\fS--dry-run\fP` to make dry runs.

.PP
This package is based on the <Bash Library>.

.SH OPTIONS

.PP
\fIThe following common options are supported:\fP
.TP
\fB-p | --project\fP =path
define the project directory path (default is `\fSpwd\fP` - `\fSPATH\fP` must exist)
.TP
\fB-d | --working-dir\fP =path
redefine the working directory (default is `\fSpwd\fP` - `\fSPATH\fP` must exist)
.TP
\fB-h | --help\fP
show this information message 
.TP
\fB-v | --verbose\fP
increase script verbosity 
.TP
\fB-q | --quiet\fP
decrease script verbosity, nothing will be written unless errors 
.TP
\fB-f | --force\fP
force some commands to not prompt confirmation 
.TP
\fB-i | --interactive\fP
ask for confirmation before any action 
.TP
\fB-x | --debug | --dry-run\fP
see commands to run but do not run them actually
.PP
You can group short options like `\fS-xc\fP`, set an option argument like `\fS-d(=)value\fP` or
`\fS--long=value\fP` and use `\fS--\fP` to explicitly specify the end of the script options.

.PP
\fIThe following actions are currently available:\fP

.SS cleanup

.PP
This will clean (remove) all OS or IDE specific files from the project
(configuration variable: `\fSDEFAULT_CLEANUP_NAMES\fP`).

.RS
.IP \fBSyntax\fP

.PP
dev-tools.sh  -[\fIcommon options\fP ...]  [\fB--dry-run\fP]  \fBcleanup\fP  --

.RS
.RS
.SS config

.PP
Manage the dev-tools configuration for a package, stored in `\fS.devtools\fP` dotfile ;
with no option, current configuration will be shown.

.RS
.IP \fBSyntax\fP

.PP
dev-tools.sh  -[\fIcommon options\fP ...]  [\fB--dry-run\fP]
    [\fB--var\fP \fI=name\fP]  [\fB--val\fP \fI=value\fP]  [\fB--filename\fP]  [\fB--full\fP]  \fBconfig\fP  --

.RE
.RS
.IP \fBOptions\fP
.TP
\fB--var\fP =name
select a configuration variable to read or define
.TP
\fB--val\fP =value
define a configuration variable value (requires the `\fS--var\fP` option to be defined)
.TP
\fB--filename\fP
see current configuration file path for the project
.TP
\fB--full\fP
see the full configuration entries for the project (defaults and custom)
.RS
.RS
.SS deploy

.PP
Will search for files suffixed by `\fS__ENV__\fP` in the project path (recursively) and
over-write the original ones (without suffix).

.RS
.IP \fBSyntax\fP

.PP
dev-tools.sh  -[\fIcommon options\fP ...]  [\fB--dry-run\fP]
    [\fB--env\fP \fI=env\fP]  \fBdeploy\fP  --

.RE
.RS
.IP \fBOptions\fP
.TP
\fB--env\fP =name
the environment shortcut to deploy (default is `\fSDEFAULT\fP` - configuration variable: `\fSDEFAULT_DEPLOY_ENV\fP`)
.RS
.RS
.SS extract

.PP
Will search and extract strings from files contents recursively ; result is written on STDOUT
but can be stored in a file.

.RS
.IP \fBSyntax\fP

.PP
dev-tools.sh  -[\fIcommon options\fP ...]  [\fB--dry-run\fP]
    [\fB--begin\fP \fI=mask\fP]  [\fB--end\fP \fI=mask\fP]  [\fB--output\fP \fI=filename\fP]
    [\fB--filename\fP]  \fBextract\fP  --

.RE
.RS
.IP \fBOptions\fP
.TP
\fB--begin\fP =mask
the mask to use to begin the matching (configuration variable: `\fSDEFAULT_EXTRACT_BEGIN_MASK\fP`) 
.TP
\fB--end\fP =mask
the mask to use to end the matching (configuration variable: `\fSDEFAULT_EXTRACT_END_MASK\fP`) 
.TP
\fB--output\fP =filename
a filename to write results in (this will overwrite any existing file)
.TP
\fB--show-filename\fP
write matching filenames before extracted content (configuration variable: `\fSDEFAULT_EXTRACT_SHOW_FILENAME\fP`)
.RS
.RS
.SS fix-rights

.PP
This will fix files and directories UNIX rights recursively on the project.

.RS
.IP \fBSyntax\fP

.PP
dev-tools.sh  -[\fIcommon options\fP ...]  [\fB--dry-run\fP]
    [\fB--files\fP \fI=chmod\fP]  [\fB--dirs\fP \fI=chmod\fP]  [\fB--bin\fP \fI=path\fP]
    [\fB--bin-mask\fP \fI=mask\fP]  \fBfix-rights\fP  --

.RE
.RS
.IP \fBOptions\fP
.TP
\fB--dirs\fP =chmod
the rights level setted for directories (default is `\fS0755\fP` - configuration variable: `\fSDEFAULT_FIXRIGHTS_DIRS_CHMOD\fP`) 
.TP
\fB--files\fP =chmod
the rights level setted for files (default is `\fS0644\fP` - configuration variable: `\fSDEFAULT_FIXRIGHTS_FILES_CHMOD\fP`) 
.TP
\fB--bin\fP =path
directory name of the binaries, to define their rights on `\fSa+x\fP` (default is `\fSbin/\fP` - configuration variable: `\fSDEFAULT_FIXRIGHTS_BIN_DIR\fP`)
.TP
\fB--bin-mask\fP =mask
mask to match binary files in 'bin' (default is empty - configuration variable: `\fSDEFAULT_FIXRIGHTS_BIN_MASK\fP`)
.RS
.RS
.SS manpage

.PP
Build a manpage file based on a markdown content ; the manpage is added in system manpages
and can be referenced if the `\fSwhatis\fP` and `\fSmakewhatis\fP` binaries are found or defined.

.RS
.IP \fBSyntax\fP

.PP
dev-tools.sh  -[\fIcommon options\fP ...]  [\fB--dry-run\fP]
    [\fB--source\fP \fI=path\fP]  [\fB--filename\fP \fI=filename\fP]  [\fB--type\fP \fI=type\fP]  [\fB--dir\fP \fI=path\fP]
    [\fB--markdown\fP \fI=bin\fP]  [\fB--whatis\fP \fI=bin\fP]  [\fB--makewhatis\fP \fI=bin\fP]  \fBmanpage\fP  --

.RE
.RS
.IP \fBOptions\fP
.TP
\fB--source\fP =filename
the manpage source file (default is `\fSMANPAGE.md\fP` - configuration variable: `\fSDEFAULT_MANPAGE_SOURCE\fP`) 
.TP
\fB--filename\fP =filename
the filename to use to create the manpage (configuration variable: `\fSDEFAULT_MANPAGE_FILENAME\fP`) 
.TP
\fB--section\fP =reference
the manpage section (default is '3' - configuration variable: `\fSDEFAULT_MANPAGE_SECTION\fP`) 
.TP
\fB--dir\fP =name
the manpage system directory to install manpage in 
.TP
\fB--markdown\fP =bin path
the binary to use for the 'markdown' command 
(default is installed MarkdownExtended package - configuration variable: `\fSDEFAULT_MANPAGE_MARKDOWN_BIN\fP`) 
.TP
\fB--whatis\fP =bin path
the binary to use for the 'whatis' command (configuration variable: `\fSDEFAULT_MANPAGE_WHATIS_BIN\fP`) 
.TP
\fB--makewhatis\fP =bin path
the binary to use for the 'makewhatis' command (configuration variable: `\fSDEFAULT_MANPAGE_MAKEWHATIS_BIN\fP`)
.RS
.RS
.SS sync

.PP
Will `\fSrsync\fP` a project directory to a target, which can use SSH protocol if so ; use the
`\fS-x\fP` option to process a `\fS--dry-run\fP` rsync.

.RS
.IP \fBSyntax\fP

.PP
dev-tools.sh  -[\fIcommon options\fP ...]  [\fB--dry-run\fP]
    [\fB--env\fP \fI=env\fP]  [\fB--target\fP \fI=server\fP]  [\fB--options\fP \fI="rsync options"\fP]  \fBsync\fP  --

.RE
.RS
.IP \fBOptions\fP
.TP
\fB--target\fP =server
the server name to use for synchronization (configuration variable: `\fSDEFAULT_SYNC_SERVER\fP`) 
.TP
\fB--options\fP ="rsync opts"
an options string used for the 'rsync' command (configuration variable: `\fSDEFAULT_SYNC_RSYNC_OPTIONS\fP`) 
.TP
\fB--env\fP =env
the environment shortcut to deploy if so (configuration variable: `\fSDEFAULT_SYNC_ENV\fP`)
.RS
.RS
.SS version-tag

.PP
This will create a new GIT version TAG according to the semantic versioning (see <http://semver.org/>).

.RS
.IP \fBSyntax\fP

.PP
dev-tools.sh  -[\fIcommon options\fP ...]  [\fB--dry-run\fP]
    [\fB--name\fP \fI=version\fP]  [\fB--branch\fP \fI=name\fP]  [\fB--hook\fP \fI=path\fP]  [\fB--no-hook\fP]  \fBversion-tag\fP  --

.RE
.RS
.IP \fBOptions\fP
.TP
\fB--name\fP =version
the name of the new tag ; default will be next increased version number 
.TP
\fB--branch\fP =name
which branch to use (default is `\fSmaster\fP` - configuration variable: `\fSDEFAULT_VERSIONTAG_BRANCH\fP`)
.TP
\fB--hook\fP =path
define a pre-tag hook file (configuration variable: `\fSDEFAULT_VERSIONTAG_HOOK\fP` - see `\fSpre-tag-hook.sample\fP`)
.TP
\fB--no-hook\fP
do not run any pre-tag hook file (disable config setting)
.RS
.RS
.RS
.SH ENVIRONMENT

.PP
The script doesn't really define environment variables but handles a set of configuration
variables that can be overwritten or modified to fit your needs and special environment.
If you want to define a configuration value globally, edit the `\fSdev-tools.conf\fP` file directly,
which is loaded at any call of the script. You can also define "per project" configuration
settings creating a `\fS.devtools\fP` file at the root of the project. The `\fSconfig\fP` action of
the script can help you to manage this type of configuration.

.PP
\fIThe following configuration variables are available:\fP
.TP
DEFAULT_CLEANUP_NAMES
list of file names or masks to remove when cleaning a project
.TP
DEFAULT_DEPLOY_ENV
default environment name to deploy when using action `\fSdeploy\fP`
.TP
DEFAULT_EXTRACT_BEGIN_MASK DEFAULT_EXTRACT_END_MASK
the default masks to begin and end file contents extraction when using action `\fSextract\fP`
.TP
DEFAULT_EXTRACT_SHOW_FILENAME
whether to show source filename before contents extracted when using action `\fSextract\fP`
.TP
DEFAULT_FIXRIGHTS_BIN_DIR
the default binaries path in the project when using action `\fSfix-rights\fP`
.TP
DEFAULT_FIXRIGHTS_BIN_MASK
the default mask to match binary files when using action `\fSfix-rights\fP`
.TP
DEFAULT_FIXRIGHTS_FILES_CHMOD DEFAULT_FIXRIGHTS_DIRS_CHMOD
default rights levels to use on files and directories when using action `\fSfix-rights\fP`
.TP
DEFAULT_MANPAGE_SOURCE DEFAULT_MANPAGE_FILENAME
default source and target file names when using action `\fSmanpage\fP`
.TP
DEFAULT_MANPAGE_SECTION
default system manpage type to use when using action `\fSmanpage\fP`
.TP
DEFAULT_MANPAGE_WHATIS_BIN DEFAULT_MANPAGE_MAKEWHATIS_BIN DEFAULT_MANPAGE_MARKDOWN_BIN
path of the binaries to use for the `\fSwhatis\fP`, `\fSmakewhatis\fP` and `\fSmarkdown\fP` commands
when using action `\fSmanpage\fP`
.TP
DEFAULT_SYNC_SERVER DEFAULT_SYNC_ENV
default distant server and environment to synchronize when using action `\fSsync\fP`
.TP
DEFAULT_SYNC_RSYNC_OPTIONS
default options to use with the `\fSrysnc\fP` command when using action `\fSsync\fP`
.TP
DEFAULT_VERSIONTAG_BRANCH
default branch name to use to create tags when using action `\fSversion-tag\fP`
.TP
DEFAULT_VERSIONTAG_HOOK
path of the hook filename when using action `\fSversion-tag\fP`
.SH FILES
.TP
\fBdev-tools.sh\fP
The library source file ; this is the script name to call in command line ; it can be
stored anywhere in the file system ; its relevant place could be `\fS$HOME/bin\fP` for a user
or, for a global installation, in a place like `\fS/usr/local/bin\fP` (be sure to put it in
a directory included in the global `\fS$PATH\fP`).
.TP
\fBdev-tools.conf\fP
The global script configuration file ; this file is required and will be searched in
the same directory as the script above, then in current user `\fS$HOME\fP`, then in system
configurations `\fS/etc\fP`.
.TP
\fBdev-tools-actions/\fP
This directory contains the actions currently available ; the directory and its contents
are required to use script's actions ; they will be searched in the same directory as
the script above, then in current user `\fS$HOME\fP`.
.TP
\fB.devtools\fP
This is the specific dotfile to use for "per project" configuration ; you may write your
configuration following the global `\fSdev-tools.conf\fP` rules ; this file is searched at the
root directory of each project (defined by the '-p' option).
.TP
\fBbash-library/\fP
This directory embeds the required third-party <Bash Library>.
If you already have a version of the library installed in your system, you can over-write
the library loaded (and skip the embedded version) re-defining the `\fSDEFAULT_BASHLIBRARY_PATH\fP`
of the global configuration file.
.SH EXAMPLES

.PP
A "classic" usage of the script would be:
.RS

.EX
dev-tools.sh -p ../relative/path/to/concerned/project action
.EE
.RE

.PP
To get an help string, run:
.RS

.EX
dev-tools.sh -h OR dev-tools.sh -h action
.EE
.RE

.PP
To make a dry run before really executing the actions, use:
.RS

.EX
dev-tools.sh --dry-run ... action
.EE
.RE

.SH LICENSE

.PP
The library is licensed under GPL-3.0 - Copyleft (c) Les Ateliers Pierrot
<http://www.ateliers-pierrot.fr/> - Some rights reserved. For documentation,
sources & updates, see <http://github.com/atelierspierrot/dev-tools>. 
To read GPL-3.0 license conditions, see <http://www.gnu.org/licenses/gpl-3.0.html>.

.SH BUGS

.PP
To transmit bugs, see <http://github.com/atelierspierrot/dev-tools/issues>.

.SH AUTHOR

.PP
\fBLes Ateliers Pierrot\fP <http://www.ateliers-pierrot.fr/>.

.SH SEE ALSO

.PP
bash-library(3)



dev-tools.sh
#!/bin/bash
#
# DevTools - Packages development & deployment facilities
# Copyleft (c) 2013 Pierre Cassat and contributors
# <www.ateliers-pierrot.fr> - <contact@ateliers-pierrot.fr>
# License GPL-3.0 <http://www.opensource.org/licenses/gpl-3.0.html>
# Sources <http://github.com/atelierspierrot/dev-tools>
#
# Global help : `dev-tools.sh -h`
# Action help : `dev-tools.sh -h action`
# Action usage : `dev-tools.sh [-vix] -p=PROJECT_PATH [action options] action`
#

###### First paths
declare -rx _REALPATH="$0"
declare -rx _REALDIRPATH="`dirname $_REALPATH`"
declare -rx _DEVTOOLS_CONFIGFILE="dev-tools.conf"
declare -rx _DEVTOOLS_ACTIONSDIR="dev-tools-actions"

#### findRequirements ( path , info string )
# search for a relative or root path
# this needs to be defined first as it is used to find the library
findRequirements() {
    REQFILE="$1"
    ROOTREQFILE="${_REALDIRPATH}/${REQFILE}"
    REQINFO="$2"
    if [ -e "$REQFILE" ]; then
        echo "$REQFILE"
        return 0
    elif [ -e "$ROOTREQFILE" ]; then
        echo "$ROOTREQFILE"
        return 0
    else
        PADDER=$(printf '%0.1s' "#"{1..1000})
        printf "\n### %*.*s\n    %s\n    %s\n%*.*s\n\n" 0 $(($(tput cols)-4)) "ERROR! $PADDER" \
            "Unable to find required $REQINFO '$REQFILE'!" \
            "Sent in '$0' line '${LINENO}' by '`whoami`' - pwd is '`pwd`'" \
            0 $(tput cols) "$PADDER";
        exit 1
    fi
}

######## Inclusion of the config
declare -rx CFGFILE=`findRequirements "${_DEVTOOLS_CONFIGFILE}" "configuration file"`
if [ -f "$CFGFILE" ]; then source "$CFGFILE"; else echo "$CFGFILE"; exit 1; fi

######## Inclusion of the lib
declare -rx LIBFILE=`findRequirements "${DEFAULT_BASHLIBRARY_PATH}" "bash library"`
if [ -f "$LIBFILE" ]; then source "$LIBFILE"; else echo "$LIBFILE"; exit 1; fi

######## Path of the actions
declare -rx _BASEDIR=`findRequirements "${_DEVTOOLS_ACTIONSDIR}" "actions directory"`
if [ ! -d "$_BASEDIR" ]; then echo "$_BASEDIR"; exit 1; fi

#### load_actions_infos ()
# load the available actions in $ACTIONS_LIST
# load the available actions synopsis in $ACTIONS_SYNOPSIS
# load the available actions synopsis in $ACTIONS_DESCRIPTIONS
# this needs to be defined here as it is used for strings below
load_actions_infos () {
    ACTIONS_LIST=()
    ACTIONS_SYNOPSIS=()
    ACTIONS_DESCRIPTIONS=()
    export SCRIPTMAN=true
    for action in $_BASEDIR/*.sh; do
        ACTION_SYNOPSIS=''
        ACTION_DESCRIPTION=''
        ACTION_LONGDESCRIPTION=''
        ACTION_CFGVARS=''
        myaction=${action##$_BASEDIR/}
        pos=${#ACTIONS_LIST[@]}
        source "${action}"
        ACTIONS_LIST[$pos]="${myaction%%.sh}"
        if [ -n "$ACTION_SYNOPSIS" ]; then
            ACTIONS_SYNOPSIS[$pos]="${ACTION_SYNOPSIS}"
        fi
        if [ -n "$ACTION_DESCRIPTION" ]; then
            ACTIONS_DESCRIPTIONS[$pos]="${ACTION_DESCRIPTION}"
        fi
        if [ -n "$ACTION_CFGVARS" ]; then
            ACTIONS_CFGVARS=("${ACTIONS_CFGVARS[@]}" "${ACTION_CFGVARS[@]}")
        fi
    done
    export ACTIONS_LIST ACTIONS_SYNOPSIS ACTIONS_DESCRIPTIONS ACTIONS_CFGVARS
    export SCRIPTMAN=false
}

#### script settings ##########################

# bash-lib settings
MANPAGE_NODEPEDENCY=true
COMMON_OPTIONS_ARGS="p:${COMMON_OPTIONS_ARGS}"

# paths
declare -x _BACKUP_DIR="${_BASEDIR}/backup/"
declare -x _PROJECT="project"
declare -x _TARGET=`pwd`
declare -x SCRIPTMAN=false

# per action vars
declare -x ACTION
declare -x ACTION_FILE

# all actions vars
declare -xa ACTIONS_LIST=() && declare -xa ACTIONS_SYNOPSIS=() && declare -xa ACTIONS_DESCRIPTIONS=() &&
    declare -xa ACTIONS_CFGVARS=( DEFAULT_CONFIG_FILE DEFAULT_BASHLIBRARY_PATH ) && \
    load_actions_infos;

# script infos
declare -x NAME="DevTools"
declare -x PRESENTATION="Packages development & deployment facilities"
declare -x SYNOPSIS="$LIB_SYNOPSIS_ACTION"
declare -x DEPLOY_HELP="Run option '-h' for help.";
declare -x DEPLOY_ACTIONS_HELP="Run option '-h action' for help about a specific action.";
declare -x SHORT_DESCRIPTION="This helper script will assist you in creating version tags of a git repository, deploying a project and its environment dependencies etc.\n\
\tRun option '<bold>-h action</bold>' to see the help about a specific action and use option '<bold>--dry-run</bold>' to make dry runs.";
declare -x SEE_ALSO="This tool is an open source stuff licensed under GNU/GPL v3: <http://github.com/atelierspierrot/dev-tools>\n\
\tTo transmit a bug or an evolution: <http://github.com/atelierspierrot/dev-tools/issues>\n\
\tThis tool is base on the Bash Library: <http://github.com/atelierspierrot/bash-library>"

# actions infos
declare -rx ACTION_PRESENTATION_MASK="Help for action \"<bold>%s</bold>\"";
declare -rx ACTION_SYNOPSIS_MASK="~\$ <bold>${0}</bold>  -[<underline>common options</underline>]  %s  %s  --";
actionsstr=""
actionsdescription=""
actionssynopsis=""
for i in ${!ACTIONS_LIST[*]}; do
    itemstr=${ACTIONS_LIST[$i]}
    if [ "${#actionsstr}" == 0 ]; then
        actionsstr="${itemstr}"
    else
        actionsstr="${actionsstr} | ${itemstr}"
    fi
    actionsdescription="${actionsdescription}\n\n\t<bold>${itemstr}</bold>\n"
    itemdesc=${ACTIONS_DESCRIPTIONS[$i]}
    if [ -n "${itemdesc}" ]; then
        actionsdescription="${actionsdescription}\t${itemdesc}";
    fi
    itemsyn=${ACTIONS_SYNOPSIS[$i]}
    if [ -n "${itemsyn}" ]; then
        actionsdescription="${actionsdescription}\n\t`printf \"${ACTION_SYNOPSIS_MASK}\" \"${itemsyn}\" \"${itemstr}\"`";
        actionssynopsis="${actionssynopsis}\n\t... ${itemsyn}  ${itemstr}"
    else
        actionsdescription="${actionsdescription}\n\t`printf \"${ACTION_SYNOPSIS_MASK}\" '' \"${itemstr}\"`";
    fi
done
declare -x DESCRIPTION="${SHORT_DESCRIPTION}\n\n<bold>AVAILABLE ACTIONS</bold>${actionsdescription}"
declare -x OPTIONS="Below is a list of common options available ; each action can accepts other options.\n\n\
\t<bold>-p | --project=PATH</bold>\tthe project path (default is 'pwd' - 'PATH' must exist)\n\
\t<bold>-d | --working-dir=PATH</bold>\tredefine the working directory (default is 'pwd' - 'PATH' must exist)\n\
\t<bold>-h | --help</bold>\t\tshow this information message \n\
\t<bold>-v | --verbose</bold>\t\tincrease script verbosity \n\
\t<bold>-q | --quiet</bold>\t\tdecrease script verbosity, nothing will be written unless errors \n\
\t<bold>-f | --force</bold>\t\tforce some commands to not prompt confirmation \n\
\t<bold>-i | --interactive</bold>\task for confirmation before any action \n\
\t<bold>-x | --debug</bold>\t\tsee commands to run but not run them actually\n\
\t<bold>--dry-run</bold>\t\talias of '-x'\n\
\n${OPTIONS_USAGE_INFOS}";
declare -x SYNOPSIS_ERROR="${0}  [-${COMMON_OPTIONS_ARGS}]  [-x|--dry-run]  ... ${actionssynopsis}\n\
\t-p | --project=path  <action : ${actionsstr}> --";

#### internal lib ##########################

#### action_file ( action name )
# find action file
action_file () {
    local ACTION="$1"
    if [ ! -z "$ACTION" ]; then
        ACTION_FILE="${_BASEDIR}/${ACTION}.sh"
        # hack to allow direct call of a file as action (with path from root)
        if [ ! -f "$ACTIONFILE" ]; then
            TMP_ACTIONFILE="${_REALDIRPATH}/${ACTION}"
            if [ -f "$TMP_ACTIONFILE" ]; then
                ACTION_FILE="$TMP_ACTIONFILE"
            fi
        fi    
    fi
    export ACTION_FILE
}

#### action_usage ( action name, action file )
# usage string per action
action_usage () {
    local ACTION_NAME="$1"
    local ACTION_FILE="$2"
    export SCRIPTMAN=true
    if [ -f $ACTION_FILE ]; then
        ACTION_SYNOPSIS=''
        ACTION_DESCRIPTION=''
        ACTION_OPTIONS=''
        ACTION_CFGVARS=''
        source "$ACTION_FILE"
        local TMP_USAGE="\n<bold>NAME</bold>\n\
\t`printf \"${ACTION_PRESENTATION_MASK}\" \"${ACTION_NAME}\"`\n\
\t${NAME}\n\n\
<bold>SYNOPSIS</bold>\n\
\t`printf \"${ACTION_SYNOPSIS_MASK}\" \"${ACTION_SYNOPSIS}\" \"${ACTION_NAME}\"`";
        if [ -n "${ACTION_DESCRIPTION}" ]; then
            TMP_USAGE="${TMP_USAGE}\n\n<bold>DESCRIIPTION</bold>\n\t${ACTION_DESCRIPTION}";
        fi
        if [ -n "${ACTION_OPTIONS}" ]; then
            TMP_USAGE="${TMP_USAGE}\n\n<bold>OPTIONS</bold>\n\t${ACTION_OPTIONS}";
        fi
        if [ -n "${ACTION_CFGVARS}" ]; then
            TMP_USAGE="${TMP_USAGE}\n\n<bold>ENVIRONMENT</bold>\n\tAvailable configuration variables: <${COLOR_NOTICE}>${ACTION_CFGVARS[@]}</${COLOR_NOTICE}>";
        fi
        if [ -n "${ACTION_FILE}" ]; then
            TMP_USAGE="${TMP_USAGE}\n\n<bold>FILE</bold>\n\t<underline>${ACTION_FILE}</underline>";
        fi
        local TMP_VERS="`library_info`"
        TMP_USAGE="${TMP_USAGE}\n\n<${COLOR_COMMENT}>${TMP_VERS}</${COLOR_COMMENT}>";
        parsecolortags "$TMP_USAGE"
    else
        error "Action file $ACTION_FILE not found!"
    fi
    export SCRIPTMAN=false
    return 0;
}

#### root_required ()
# ensure current user is root
root_required () {
    THISUSER=`whoami`
    if [ "$THISUSER" != "root" ]; then
        error "You need to run this as a 'sudo' user!"
    fi
}

#### targetdir_required ()
# ensure the project root directory is defined
targetdir_required () {
    if [ -z "$_TARGET" ]; then
        prompt 'Target directory of the project to work on' `pwd` ''
        export _TARGET=$USERRESPONSE
    fi
    if [ ! -d "$_TARGET" ]; then error "Unknown root directory '${_TARGET}' !"; fi
    load_target_config
}

#### backupdir_required ()
# ensure the project backup directory exists
backupdir_required () {
    if [ ! -d "${_BACKUP_DIR}" ]; then
        mkdir "${_BACKUP_DIR}" && chmod 775 "${_BACKUP_DIR}"
        if [ ! -d "${_BACKUP_DIR}" ]; then
            error "Backup directory '${_BACKUP_DIR}' can't be created (try to run this script as 'sudo') !"
        fi
    fi
}

#### trigger_event ( command )
## trigger a user defined command for an event
trigger_event () {
    iexec "$1"
}

#### load_target_config ()
# overwrite current deploy config with target's custom one if so
load_target_config () {
    target_configfile=$(realpath "${_TARGET}/${DEFAULT_CONFIG_FILE}")
    if [ -f "$target_configfile" ]; then
        source "$target_configfile"
    fi
    for p in "${ACTIONS_CFGVARS[@]}"; do
        export "$p"
    done
}

#### parsecomonoptions ( "$@" )
## over-writing of default function
parsecomonoptions () {
    local oldoptind=$OPTIND
    local options=$(getscriptoptions "$@")
    export ACTION=$(getlastargument $options)
    if [ ! -z $ACTION ]; then
        action_file $ACTION
        if [ ! -f "$ACTION_FILE" ]; then
            simple_error "unknown action '${ACTION}'"
        fi
    fi
    while getopts ":p:${COMMON_OPTIONS_ARGS}" OPTION $options; do
        OPTARG="${OPTARG#=}"
        case $OPTION in
        # common options
            h) 
                if [ -z $ACTION ]
                    then usage;
                    else action_usage $ACTION $ACTION_FILE;
                fi
                exit 0;;
            i) export INTERACTIVE=true; export QUIET=false;;
            v) export VERBOSE=true; export QUIET=false;;
            f) export FORCED=true;;
            x) export DEBUG=true; verecho "- debug option enabled: commands shown as 'debug >> \"cmd\"' are not executed";;
            q) export VERBOSE=false; export INTERACTIVE=false; export QUIET=true;;
            d) setworkingdir $OPTARG;;
            l) setlogfilename $OPTARG;;
            V) script_version; exit 0;;
            p) export _TARGET=$OPTARG;;
            -) case $OPTARG in
        # common options
                    project*) export _TARGET=$LONGOPTARG;;
                    help|man|usage) 
                        if [ -z $ACTION ]
                            then usage;
                            else action_usage $ACTION $ACTION_FILE;
                        fi
                        exit 0;;
                    vers*) script_version; exit 0;;
                    interactive) export INTERACTIVE=true; export QUIET=false;;
                    verbose) export VERBOSE=true; export QUIET=false;;
                    force) export FORCED=true;;
                    debug | dry-run*) export DEBUG=true; verecho "- debug option enabled: commands shown as 'debug >> \"cmd\"' are not executed";;
                    quiet) export VERBOSE=false; export INTERACTIVE=false; export QUIET=true;;
                    working-dir*) LONGOPTARG="`getlongoptionarg \"${OPTARG}\"`"; setworkingdir $LONGOPTARG;;
                    log*) LONGOPTARG="`getlongoptionarg \"${OPTARG}\"`"; setlogfilename $LONGOPTARG;;
        # library options
                    libhelp) clear; library_usage; exit 0;;
                    libvers*) library_version; exit 0;;
                    libdoc*) libdoc; exit 0;;
        # no error for others
                    *) rien=rien;;
                esac ;;
            \?) rien=rien;;
        esac
    done
    export OPTIND=$oldoptind
    return 0
}

#### first setup & options treatment ##########################

parsecomonoptions "$@"

if [ ! -z "$ACTION" ]
then
    PREACTION="EVENT_PRE_${ACTION}"
    POSTACTION="EVENT_POST_${ACTION}"
    # executing requested action
    if [ -f "$ACTION_FILE" ]
    then
        export _BASEDIR _BACKUP_DIR _PROJECT _TARGET
        if [ ! -z "${!PREACTION}" ]; then
            trigger_event "${!PREACTION}"
        fi
        source "$ACTION_FILE"
        if [ ! -z "${!POSTACTION}" ]; then
            trigger_event "${!POSTACTION}"
        fi
    else
        simple_error "unknown action '${ACTION}'"
    fi
else
    simple_error "no action to execute"
fi

exit 0
# Endfile
